lilya/_internal/_encoders2.py

```diff
@@ -0,0 +1,272 @@
+from __future__ import annotations
+
+import json
+from collections import deque
+from collections.abc import Callable, Generator, Iterable, Sequence
+from contextvars import ContextVar
+from dataclasses import asdict, is_dataclass
+from datetime import date, datetime
+from enum import Enum
+from inspect import isclass
+from pathlib import PurePath
+from typing import Any, Protocol, TypeVar, cast, runtime_checkable
+
+T = TypeVar("T")
+
+
+# TODO: timedelta encode
+@runtime_checkable
+class EncoderProtocol(Protocol[T]):
+    def is_type(self, value: Any) -> bool:
+        """Check if encoder is applicable for this type"""
+
+    def serialize(self, value: T) -> Any:
+        """ "Prepare for serialization."""
+
+
+# not runtime-checkable
+class WithEncodeProtocol(Protocol):
+    def is_type_structure(self, value: Any) -> bool:
+        """Check if encoder is applicable for this type"""
+
+    def encode(self, structure: Any, value: Any) -> Any:
+        """
+        Function that transforms a value into the structure
+        """
+
+
+class Encoder(EncoderProtocol[T]):
+    """
+    The base class for any custom encoder
+    added to the system.
+    """
+
+    name: str | None = None  # type: ignore
+    __type__: type | tuple[type] | None = None
+
+    def is_type(self, value: Any) -> bool:
+        return isinstance(value, self.__type__)
+
+    def is_type_structure(self, value: Any) -> bool:
+        return issubclass(value, self.__type__)
+
+
+class DataclassEncoder(EncoderProtocol, WithEncodeProtocol):
+    name: str = "DataclassEncoder"
+
+    def is_type(self, value: Any) -> bool:
+        return cast(bool, is_dataclass(value))
+
+    is_type_structure = is_type
+
+    def serialize(self, obj: Any) -> Any:
+        return asdict(obj)
+
+    def encode(self, structure: Any, value: Any) -> Any:
+        return structure(**value)
+
+
+class NamedTupleEncoder(EncoderProtocol, WithEncodeProtocol):
+    name: str = "NamedTupleEncoder"
+
+    def is_type(self, value: Any) -> bool:
+        return isinstance(value, tuple) and hasattr(value, "_asdict")
+
+    def is_type_structure(self, value: Any) -> bool:
+        return issubclass(value, tuple) and hasattr(value, "_asdict")
+
+    def serialize(self, obj: Any) -> dict:
+        return cast(dict, obj._asdict())
+
+    def encode(self, structure: type[Any], obj: Any) -> Any:
+        if isinstance(obj, dict):
+            return structure(**obj)
+        return structure(*obj)
+
+
+class ModelDumpEncoder(EncoderProtocol, WithEncodeProtocol):
+    name: str = "ModelDumpEncoder"
+    # e.g. pydantic
+
+    def is_type(self, value: Any) -> bool:
```

Comment:
I can easily bypass this. I have done something in the past and I used the name model_dump a lot before pydantic 2 came out. 

If I use this here, it will be considered s model dump encoder? That means it could be pydantic or anything else? 

Ok, I like the approach.